lib/auth/auth_with_roles.go

```diff
@@ -1167,21 +1167,25 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 	var roles []string
 	var traits wrappers.Traits
 
-	switch {
-	case a.hasBuiltinRole(string(teleport.RoleAdmin)):
-		// If it's an admin generating the certificate, the roles and traits for
-		// the user have to be fetched from the backend. This should be safe since
-		// this is typically done against a local user.
-		user, err := a.GetUser(req.Username, false)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		roles = user.GetRoles()
-		traits = user.GetTraits()
-	case req.Username == a.context.User.GetName():
-		// user is requesting TTL for themselves,
-		// limit the TTL to the duration of the session, to prevent
-		// users renewing their certificates forever
+	// this prevents clients who have no chance at getting a cert and impersonating anyone
+	// from enumerating local users and hitting database
+	if !a.hasBuiltinRole(string(teleport.RoleAdmin)) && !a.context.Checker.CanImpersonateSomeone() && req.Username != a.context.User.GetName() {
+		return nil, trace.AccessDenied("access denied")
+	}
+
+	// Extract the user and role set for whom the certificate will be generated.
+	// This should be safe since
+	// this is typically done against a local user.
```

Comment:
> this is typically done against a local user.

Can we make this a hard rule?  Allowing impersonation of users who are managed by external identity providers feels like it would allow for some scenarios that are fairly difficult to reason about.  If we can disallow it, I think we should.  Something like:

```golang
if user.GetCreatedBy().Connector != nil {
    log.Warningf("User %v tried to issue a cert for externally managed user %v, this is not supported.", a.context.User.GetName(), req.Username)
    return trace.AccessDeneid("access denied")
}
```